pubsub/producer.go

```diff
@@ -0,0 +1,331 @@
+	}
+	for k, v := range acked {
+		// Only re-insert messages that were removed the the pending list first.
+		_, err := p.reproduce(ctx, v, k)
+		if err != nil {
+			log.Error("Re-inserting pending messages with inactive consumers", "error", err)
+		}
+	}
+	return p.cfg.CheckPendingInterval
+}
+
+// checkResponses checks iteratively whether response for the promise is ready.
+func (p *Producer[Request, Response]) checkResponses(ctx context.Context) time.Duration {
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	for id, promise := range p.promises {
+		res, err := p.client.Get(ctx, id).Result()
+		if err != nil {
+			if errors.Is(err, redis.Nil) {
+				continue
+			}
+			log.Error("Error reading value in redis", "key", id, "error", err)
+		}
+		val, err := p.mResp.Unmarshal([]byte(res))
+		if err != nil {
+			log.Error("Error unmarshaling", "value", res, "error", err)
+			continue
+		}
+		promise.Produce(val)
+		delete(p.promises, id)
+	}
+	return p.cfg.CheckResultInterval
+}
+
+func (p *Producer[Request, Response]) Start(ctx context.Context) {
+	p.StopWaiter.Start(ctx, p)
+}
+
+func (p *Producer[Request, Response]) promisesLen() int {
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	return len(p.promises)
+}
+
+// reproduce is used when Producer claims ownership on the pending
+// message that was sent to inactive consumer and reinserts it into the stream,
+// so that seamlessly return the answer in the same promise.
+func (p *Producer[Request, Response]) reproduce(ctx context.Context, value Request, oldKey string) (*containers.Promise[Response], error) {
+	id, err := p.client.XAdd(ctx, &redis.XAddArgs{
+		Stream: p.cfg.RedisStream,
+		Values: map[string]any{messageKey: p.mReq.Marshal(value)},
+	}).Result()
+	if err != nil {
+		return nil, fmt.Errorf("adding values to redis: %w", err)
+	}
+	p.promisesLock.Lock()
+	defer p.promisesLock.Unlock()
+	promise := p.promises[oldKey]
+	if oldKey != "" && promise == nil {
+		// This will happen if the old consumer became inactive but then ack_d
+		// the message afterwards.
+		return nil, fmt.Errorf("error reproducing the message, could not find existing one")
+	}
+	if oldKey == "" || promise == nil {
+		pr := containers.NewPromise[Response](nil)
+		promise = &pr
+	}
+	delete(p.promises, oldKey)
+	p.promises[id] = promise
+	return promise, nil
+}
+
+func (p *Producer[Request, Response]) Produce(ctx context.Context, value Request) (*containers.Promise[Response], error) {
+	p.once.Do(func() {
+		p.StopWaiter.CallIteratively(p.checkAndReproduce)
+		p.StopWaiter.CallIteratively(p.checkResponses)
+	})
+	return p.reproduce(ctx, value, "")
+}
+
+// Check if a consumer is with specified ID is alive.
+func (p *Producer[Request, Response]) isConsumerAlive(ctx context.Context, consumerID string) bool {
+	val, err := p.client.Get(ctx, heartBeatKey(consumerID)).Int64()
+	if err != nil {
+		return false
+	}
+	return time.Now().UnixMilli()-val < int64(p.cfg.KeepAliveTimeout.Milliseconds())
```

Comment:
redis timeout will delete the value and that is good enough for us.
In the race condition that we detect consumer death only on the next loop - it's o.k. to be a little late
on the other hand, we don't want to mistakingly consider a server dead because there's some offset between it's local time and our local time.

Reply:
Done.